### PR TITLE
[IE CLDNN] Permute fused ops support

### PR DESCRIPTION
**Pull request by ldebski**
[IE CLDNN] Addition of permute fused ops support and tests for this situation. This fusion can be perform with simple primitives (activation, quantization, eltwise, scale).

JIRA: CVS-29383

### CODE REVIEW

**Comment by lznamens** [comment]
You should change year in license info (2020) in all files.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 1]
```diff
@@ -3564,3 +3565,374 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, deconv_scale_actv_quant_u8_eltw_scale_actv_
...
+                            // quantize f16 fusing not supported
+                            permute_params{CASE_PERMUTE_F16_3D_0, 6, 9},
+                            permute_params{CASE_PERMUTE_F16_3D_1, 6, 9},
+                            permute_params{CASE_PERMUTE_F16_3D_2, 6, 9},
+                            permute_params{CASE_PERMUTE_F16_3D_3, 6, 9},
+                            permute_params{CASE_PERMUTE_F16_3D_4, 6, 9},
+                        }), );
+
+
+
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3865]: nit: too many spaces

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/permute_ref.cl [hunk 2]
```diff
@@ -14,24 +14,37 @@
 
 #include "include/include_all.cl"
 
-KERNEL (permute_ref)(const __global UNIT_TYPE* input, __global UNIT_TYPE* output)
+KERNEL (permute_ref)(
+    const __global UNIT_TYPE* input,
+    __global UNIT_TYPE* output
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/permute_ref.cl]: Done

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 3]
```diff
@@ -3564,3 +3565,374 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, deconv_scale_actv_quant_u8_eltw_scale_actv_
...
+        network network_not_fused(this->engine, this->topology_non_fused, bo_not_fused);
+        network network_fused(this->engine, this->topology_fused, bo_fused);
+        network_fused.set_input_data("input", input_prim);
+        network_not_fused.set_input_data("input", input_prim);
+
+        compare(network_not_fused, network_fused, p);
+    }
+
+    layout get_input_layout(permute_params& p) {
+        // auto pad = tensor{0}.negate();
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Done

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 4]
```diff
@@ -3564,3 +3565,374 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, deconv_scale_actv_quant_u8_eltw_scale_actv_
...
+            data("eltwise_data", get_mem(layout{ p.data_type, p.input_format, p.out_shape})),
+            data("scale_data", get_mem(get_per_channel_layout(p), 5e-1f)),
+            permute("permute", "input", p.permute_order),
+            scale("scale", "permute", "scale_data"),
+            activation("actv", "scale", activation_func::relu),
+            eltwise("eltwise", "actv", "eltwise_data", eltwise_mode::sum),
+            reorder("reorder_bfyx", "eltwise", p.default_format, p.default_type)
+        );
+
+        tolerance = 1.f;
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Done

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 5]
```diff
@@ -3564,3 +3565,374 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, deconv_scale_actv_quant_u8_eltw_scale_actv_
...
+
+#define CASE_PERMUTE_F16_0 {1, 16, 4, 5}, {1, 16, 4, 5}, {0, 1, 2, 3}, tensor{0}, data_types::f16, format::b_fs_yx_fsv16, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_1 {2, 16, 4, 5}, {16, 4, 5, 2}, {1, 2, 3, 0}, tensor{0}, data_types::f16, format::b_fs_yx_fsv16, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_2 {1, 32, 2, 3}, {2, 3, 32, 1}, {2, 3, 1, 0}, tensor{0}, data_types::f16, format::b_fs_yx_fsv16, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_3 {3, 16, 1, 1}, {1, 1, 16, 3}, {3, 2, 1, 0}, tensor{0}, data_types::f16, format::b_fs_yx_fsv16, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_4 {2, 15, 4, 5}, {4, 2, 5, 15}, {2, 0, 3, 1}, tensor{0}, data_types::f16, format::bfyx, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_5 {1, 15, 1, 2}, {15, 2, 1, 1}, {1, 3, 2, 0}, tensor{0}, data_types::f16, format::bfyx, data_types::f32, format::bfyx
+#define CASE_PERMUTE_F16_6 {1, 15, 4, 4}, {4, 4, 1, 15}, {2, 3, 0, 1}, tensor{0}, data_types::f16, format::bfyx, data_types::f32, format::bfyx
+
+#define CASE_PERMUTE_i8_0 {1, 15, 4, 5}, {1, 15, 4, 5}, {0, 1, 2, 3}, tensor{0}, data_types::i8, format::bfyx, data_types::f32, format::bfyx
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Done

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 1 above)
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Done

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 6]
```diff
@@ -462,6 +466,9 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
                             input_data.get_dependency(0).get_output_layout().data_type == data_types::i8 ||
                             input_data.get_output_layout().data_type == out_layout.data_type);
 
+            should_fuse |= input_data.is_type<permute>() &&
+                            quantize_node.get_scale_shift_opt();
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: Done

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 7]
```diff
@@ -100,6 +103,7 @@ class BaseFusingTest : public ::testing::TestWithParam<T> {
 
     void SetUp() override {
         bo_fused.set_option(build_option::optimize_data(true));
+        bo_fused.set_option(build_option::graph_dumps_dir("graph_dumps/"));
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:106]: Remove debug code here and above

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 8]
```diff
@@ -3564,3 +3568,360 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, deconv_scale_actv_quant_u8_eltw_scale_actv_
...
+        activation("actv1", "scale1", activation_func::relu),
+        quantize("quant1", "actv1", "in1_lo", "in1_hi", "out1_lo", "out1_hi", 256, data_types::f32),
+        eltwise("eltw", {"quant1", "eltw_data"}, eltwise_mode::sum, p.data_type),
+        scale("scale2", "eltw", "scale2_data"),
+        activation("actv2", "scale2", activation_func::relu),
+        quantize("quant2", "actv2", "in2_lo", "in2_hi", "out2_lo", "out2_hi", 255, data_types::i8),
+        reorder("out", "quant2", p.default_format, p.default_type)
+    );
+
+    tolerance = 1e-5f;
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:3817]: I'd suggest removing `permute_quant_f32_eltwise_quant_i8` test completely and update this test to have only 1 quantize in the end of the topology (like `permute_scale_actv_eltw_scale_actv_quant_i8`) with 1.0f tolerance. 
Tests with 2 Quantize nodes fused are very sensitive to the input data and usually require tolerance ~2.0f to provide stable results. I don't think there's any difference between 1 and 2 fused quantize primitive from the Permute primitive perspective, it's more like jit generator testing which is covered by the other test cases.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 7 above)
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Removed

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 8 above)
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: I changed tests to use only single quantization with 1.0f tolerance.

**Comment by vladimir-paramuzov** [comment]
@ldebski lgtm. Please rebase and resolve conflicts.

**Pull request merged**